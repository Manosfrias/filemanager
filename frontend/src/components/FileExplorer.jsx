import React from 'react';
import PropTypes from 'prop-types';

import formatFiles from '../helpers/formatFiles';

import FileIcon from '../assets/images/file.svg';
import FolderIcon from '../assets/images/folder.svg';

import styles from "./FileExplorer.module.css";
import { useSelector } from 'react-redux';

function FileExplorer({ files = [], onClick }) {
  const formattedFiles = formatFiles(files)
  return (
    <div className={styles.FileExplorer}>
      {files.length === 0 &&
        <div className={styles.FileExplorerEmpty}>
          The root folder is empty.
        </div>
      }
      {formattedFiles && 
        <ul>
          {formattedFiles.map(file => {
            return( <li key={file.id}><FolderOrArchiver onClick={onClick} file={file} /></li>)
          })}
        </ul>
      }
    </div>
  );
}

const FolderOrArchiver = ({file, onClick}) => {
  const { selectedFile } = useSelector(store => store.filemanager);
  const handleClick = (file) => {
    onClick(file)
  }
  const selectedClass =  file.name === selectedFile?.name ? styles.FileExplorerItemSelected : ""
  return file.kind === 'folder' ? (
    <>
    <div className={`${styles.FileExplorerItem} ${selectedClass}`} key={file.id}>
      <div className={styles.FileExplorerName}>
        {file.kind === 'folder' &&
        <FolderIcon className={styles.FileExplorerIcon} />
      }
      {file.name}
      </div>
    </div>
      { file?.children.length > 0 && 
        <ul>
          {file.children.map(item => <li  key={item.id}><FolderOrArchiver onClick={onClick}  file={item} /></li>)}
        </ul>
      } 
    </>
  ) :
  (

      <button onClick={() => handleClick(file)} className={`${styles.FileExplorerItem} ${selectedClass}`} key={file.id}>
        <div className={styles.FileExplorerName}>
          <FileIcon className={styles.FileExplorerIcon} />
          {file.name}
        </div>
        
        <div className={styles.FileExplorerSize}>
          {file.size} Kb
        </div>
      </button>

  )
}

FileExplorer.propTypes = {
  files: PropTypes.arrayOf(PropTypes.object),
};

export default FileExplorer;
