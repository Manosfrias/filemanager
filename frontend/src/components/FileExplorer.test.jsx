import React from 'react';
import { render } from '@testing-library/react';

import FileExplorer from './FileExplorer';

describe('FileExplorer Component', () => {
  it('should render defaults', () => {
    const { container } = render(
      <FileExplorer />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
