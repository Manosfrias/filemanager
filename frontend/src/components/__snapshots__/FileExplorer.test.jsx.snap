// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileExplorer Component should render defaults 1`] = `
<div
  class="FileExplorer"
>
  <div
    class="FileExplorerEmpty"
  >
    The root folder is empty.
  </div>
</div>
`;
